### ISSUE TITLE
Questions of calculating the loss of padded part when model performed pre-training

### ISSUE THREAD

**Issue by kaen2891** [s3prl/s3prl#346]
Hello, I am sorry to ask you but I'm very curious about this part.

It seems to be the s3prl upstream models are calculating the padded part when performed pre-trained. For example, as following the figures:

Example 1 (Tera)
The frames around 270 ~ 320 are padded parts (for running at batch-level)
![mask_spec1](https://user-images.githubusercontent.com/46586785/170662049-5b367a3d-34ee-448f-b5c1-9bb8e82a61e2.png)
And we can know that the padded parts are weird
![pred_spec1](https://user-images.githubusercontent.com/46586785/170662083-a2596b76-6973-49f8-b8df-923297e68b74.png)
The following figure is the true (real) spectrogram
![true_spec1](https://user-images.githubusercontent.com/46586785/170662091-1c66f640-dbb4-41fd-bec7-30c55bf448a3.png)

I think this will cause makes model confusing at the pre-training level. 

Here the 2nd figures are given:

![mask_spec2](https://user-images.githubusercontent.com/46586785/170662660-14d8417e-6ba5-4a4c-bfcc-6ef52c999a39.png)
![pred_spec2](https://user-images.githubusercontent.com/46586785/170662672-937c12b5-47f0-4114-a672-51c0a7f8bdec.png)
![true_spec2](https://user-images.githubusercontent.com/46586785/170662679-311ce752-d48a-40b1-99ad-c2dfe2c6916f.png)

The order is the same as in the first figure part.

And, this leads to loss explosion like the following figure:
![image](https://user-images.githubusercontent.com/46586785/170663465-dfee4f7a-71d0-48de-96f0-caa95ad1f492.png)

If we want to prevent calculating the loss of the zero-padded part, what could we have to change the code?

And it seems that predict the 'silence' part well, but not well in the zero-padded part.

**Comment by andi611** [comment]
Following from issue https://github.com/s3prl/s3prl/issues/342,
it looks like you have modified our code, for:
1) Our [dataset](https://github.com/s3prl/s3prl/blob/master/s3prl/pretrain/bucket_dataset.py) does bucketing, so that in each batch all the utterances have the same length (for most cases, if not paddings are added, but since the bucketing was applied the paddings won't be in a long length). Thus it is not normal to have such long padding as shown in your example.
2) We calculate the actual length of each utterance without padding here: https://github.com/s3prl/s3prl/blob/6e4d75b4e149f662bb419a154109f0ac663f958d/s3prl/pretrain/mockingjay/task.py#L82 then we set those index with zero paddings to zero attention (meaning the model will not consider those time stamps): https://github.com/s3prl/s3prl/blob/6e4d75b4e149f662bb419a154109f0ac663f958d/s3prl/pretrain/mockingjay/task.py#L93-L94
3) A `mask_label` is applied, so the model will only have to predict the masked input, and certainly NOT the zero padding part. Please track the process of creating this `mask_label` here: https://github.com/s3prl/s3prl/blob/6e4d75b4e149f662bb419a154109f0ac663f958d/s3prl/pretrain/mockingjay/task.py#L155 and applying it to the loss: https://github.com/s3prl/s3prl/blob/6e4d75b4e149f662bb419a154109f0ac663f958d/s3prl/pretrain/mockingjay/pretrain_expert.py#L265

I am pretty certain our original code will NOT generate faulty examples like the ones shown above,
and it is NOT normal for the loss to explode like that.

Can you check that the above points are implemented correctly?
To me it looks like your loss is spiking mainly due to not following our implementation in 3).

FYI, I will be closing issue https://github.com/s3prl/s3prl/issues/342 since it looks like you have resolved the error,
but feel free to reopen it again if I'm mistaken. 

Andy

**Comment by kaen2891** [comment]
Dear @andi611,

Thanks for replying.

I think this problem was caused by the old version because I downloaded s3prl last October, and only modified some interesting parts.

Further, in the last issue #342, I also guess that this has occurred to the s3prl version problem.

From now, I will update all from your recent framework and try to solve this problem. We will be touch in soon.

**Comment by kaen2891** [comment]
Dear @andi611,

I have another question.

If we set the `sequence_length` from `config_model.yaml` as `1500` (default), it means the `maximum sequence length` is around 1,500 frames.

So the pre-trained model is performed until `1,500` sequences, right?

In this regard, I have some questions what if we set the maximum sequence length as 1,500.

What happens to the `3,000` frames of utterance file after the pre-training procedure in the feature extracted task or fine-tuning downstream task?

1) the given wav file will be feature extracted from the pre-trained model until 1,500 sequences.

2) the given wav file will be feature extracted from the pre-trained model until the end of that sequence?

I'm very curious about the maximum sequence length. Because I have the other datasets (not LibriSpeech), so very sensitive to set that parameter. Please help!

**Comment by andi611** [comment]
> I think this problem was caused by the old version because I downloaded s3prl last October, and only modified some interesting parts.
> Further, in the last issue #342, I also guess that this has occurred to the s3prl version problem.

I don't think this is the case...
This is our repo back from last Sept: https://github.com/s3prl/s3prl/tree/2a6894515bb7fbd9a619fba43ce035d5a8ecf401

Everything I've talked about has already been implemented.
In fact, I believe we implemented those from the very beginning.

For example:
1) The [bucketing dataset](https://github.com/s3prl/s3prl/blob/2a6894515bb7fbd9a619fba43ce035d5a8ecf401/s3prl/pretrain/mockingjay/dataset.py#L64-L66), and
2) The label mask so the model is not asked to predict the padding: https://github.com/s3prl/s3prl/blob/2a6894515bb7fbd9a619fba43ce035d5a8ecf401/s3prl/pretrain/mockingjay/pretrain_expert.py#L260
3) Calculating true length without padding: https://github.com/s3prl/s3prl/blob/2a6894515bb7fbd9a619fba43ce035d5a8ecf401/s3prl/pretrain/mockingjay/task.py#L81-L82
4) Ignore padding timestamps:
https://github.com/s3prl/s3prl/blob/2a6894515bb7fbd9a619fba43ce035d5a8ecf401/s3prl/pretrain/mockingjay/task.py#L93-L94

As such I would highly appreciate it if you could not imply that it is our older versions that caused those errors. 
Because I believe our implementations (whether old or current version) does not exhibit those faulty behaviors/errors.
Please correct me if I'm mistaken.

**Comment by andi611** [comment]
> So the pre-trained model is performed until `1,500` sequences, right?

This is partially correct. We will randomly sample a sub-sequence from the whole sequence:
https://github.com/s3prl/s3prl/blob/6e4d75b4e149f662bb419a154109f0ac663f958d/s3prl/pretrain/bucket_dataset.py#L80-L84
so it is not always performed from the start.
However, note that you can disable this any train on the whole sequence ignoring the max length.

> In this regard, I have some questions what if we set the maximum sequence length as 1,500.

We simply forward the sequence in chunks (of length determined by `sequence_length` being 1500) and concatenate the results:
https://github.com/s3prl/s3prl/blob/6e4d75b4e149f662bb419a154109f0ac663f958d/s3prl/upstream/mockingjay/builder.py#L203-L211

Please let me know if you have any more questions :)
